Sources/Linq2DynamoDb.DataContext/EntityWrapper.cs

```diff
@@ -14,6 +16,31 @@ internal class EntityWrapper : IEntityWrapper
         private readonly IEntityKeyGetter _keyGetter;
         private Document _doc, _newDoc;
 
+        private PropertyInfo _entityVersionNumberProperty;
+        private bool _hasResolvedEntityVersionNumberProperty;
```

Comment:
I'm afraid it wouldn't help either. 
Technically, what is happening here now - is that the value for _entityVersionNumberProperty is calculated per each EntityWrapper _instance_. In effect, it's recalculated all the time (because EntityWrapper instances have very short lives). Which is inefficient, and is exactly what we try to avoid throughout the library.

A typical approach to workaround this is demonstrated by [TableDefinitionWrapper.ToDocumentConversionFunctor](https://github.com/scale-tone/linq2dynamodb/blob/master/Sources/Linq2DynamoDb.DataContext/TableDefinitionWrapper.cs#L146) property. It contains a functor, that does a straightforward conversion from an entity instance to a Document. And that functor is a compiled [Expression](https://msdn.microsoft.com/en-us/library/system.linq.expressions.expression(v=vs.110).aspx), so it works as fast, as if you manually wrote an entity-specific conversion code like 

```
doc["Field1"] = entity.Field1;
doc["Field2"] = entity.Field2;
...
```

Another, less complicated, approach is used in [EntityKeyGetter](https://github.com/scale-tone/linq2dynamodb/blob/master/Sources/Linq2DynamoDb.DataContext/EntityKeyGetter.cs), where PropertyInfos are cached per TableDefinitionWrapper instance.

I would propose to use one of these two approaches, or just [Memoize ](https://github.com/scale-tone/linq2dynamodb/blob/master/Sources/Linq2DynamoDb.DataContext/Utils/GeneralUtils.cs#L105)a version value getter per entity type.